### PR TITLE
Fix: Handle aborting a job before it has started processing in Postgres

### PR DESCRIPTION
**Pull request by vchan**
If a job hasn't been dequeued yet and it is aborted, its status should be updated to `aborted`.

### CODE REVIEW

File: tests/test_queue.py [hunk 1]
```diff
@@ -461,14 +461,6 @@ async def test_job_key(self) -> None:
     async def test_schedule(self, mock_time: MagicMock) -> None:
         pass
 
-    async def test_batch(self) -> None:
```
Comment (vchan) [inline tests/test_queue.py:464]: This is already in the base test case.

File: saq/queue/postgres.py [hunk 2]
```diff
@@ -484,7 +484,29 @@ async def iter_jobs(
...
+                    jobs_table=self.jobs_table,
+                ),
+                {
+                    "key": job.key,
+                },
+            )
+            result = await cursor.fetchone()
+            await self.update(job, status=Status.ABORTING, connection=conn)
+        if result == (Status.QUEUED,):
+            await job.finish(Status.ABORTED, error=error)
```
Comment (tobymao) [inline saq/queue/postgres.py:509]: so this is all done in a transaction right? even job.finish? why didn’t you add a test? is this code tested? why did you delete one?

File: saq/queue/postgres.py [hunk 3]
```diff
@@ -484,7 +484,29 @@ async def iter_jobs(
...
+
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            await cursor.execute(
+                SQL(
+                    dedent(
+                        """
+                        SELECT status
+                        FROM {jobs_table}
+                        WHERE key = %(key)s
+                        FOR UPDATE
```
Comment (tobymao) [inline saq/queue/postgres.py:496]: do we use for update everywhere else? is this needed given the whole
block is in a transaction?

Comment (tobymao) [reply saq/queue/postgres.py:496]: why don’t we have for update in sweep?

File: saq/queue/postgres.py (hunk 2 above)
Comment (vchan) [reply saq/queue/postgres.py:509]: There's a test for abort but it didn't test this edge case. I've updated the test now. As for the deleted test, there's already a base test case that's testing batch, so it wasn't necessary. It was also wrong, which is what prompted this fix.

File: saq/queue/postgres.py (hunk 3 above)
Comment (vchan) [reply saq/queue/postgres.py:496]: I used FOR UPDATE here because if a job that hasn't been dequeued yet is aborted, I didn't want dequeue to accidentally pick it up in between the select and the update. As for sweep, do you think we should? I figured if a job's stuck or orphaned, there's no need to lock it. And if it's not, we don't want to lock them in case a worker needs to update it.

File: saq/queue/postgres.py [hunk 4]
```diff
@@ -484,7 +484,29 @@ async def iter_jobs(
...
+                    )
+                ).format(
+                    jobs_table=self.jobs_table,
+                ),
+                {
+                    "key": job.key,
+                },
+            )
+            result = await cursor.fetchone()
+            await self.update(job, status=Status.ABORTING, connection=conn)
```
Comment (tobymao) [inline saq/queue/postgres.py:507]: why isn't this in the same block? should it be

if result == Queued:
  await job.finish(Aborted):
  else:
await self.update(job, ...)

File: tests/test_queue.py [hunk 5]
```diff
@@ -452,6 +452,22 @@ async def asyncTearDown(self) -> None:
         await super().asyncTearDown()
         await teardown_postgres()
 
+    async def get_job_status(self, key: str) -> Status:
```
Comment (tobymao) [inline tests/test_queue.py:455]: make this a method in postgres? it's like the same thing right?

File: saq/queue/postgres.py [hunk 6]
```diff
@@ -484,7 +484,13 @@ async def iter_jobs(
 
     async def abort(self, job: Job, error: str, ttl: float = 5) -> None:
         job.error = error
```
Comment (tobymao) [inline saq/queue/postgres.py:486]: you don't need to set this anymore right, you can pass error to update?

Comment (vchan) [reply saq/queue/postgres.py]: You're right. Updated.
